src/libtls/tls_server.c

```diff
@@ -172,39 +181,100 @@ static bool select_suite_and_key(private_tls_server_t *this,
 			 this->server);
 		return FALSE;
 	}
-	this->suite = this->crypto->select_cipher_suite(this->crypto,
-											suites, count, key->get_type(key));
-	if (!this->suite)
-	{	/* no match for this key, try to find another type */
-		if (key->get_type(key) == KEY_ECDSA)
-		{
-			type = KEY_RSA;
-		}
-		else
-		{
-			type = KEY_ECDSA;
-		}
-		key->destroy(key);
 
-		this->suite = this->crypto->select_cipher_suite(this->crypto,
-											suites, count, type);
+	if (this->tls->get_version_max(this->tls) >= TLS_1_3)
+	{
+		/* Currently no support for derive key based on client supported
+		 * signature schemes. */
+		this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
+												  		count, KEY_ANY);
+	}
+	else
+	{
+		this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
+												  		count,
+												  		key->get_type(key));
 		if (!this->suite)
+		{	/* no match for this key, try to find another type */
+			if (key->get_type(key) == KEY_ECDSA)
+			{
+				type = KEY_RSA;
+			}
+			else
+			{
+				type = KEY_ECDSA;
+			}
+			key->destroy(key);
+
+			this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
+															count, type);
+			if (!this->suite)
+			{
+				DBG1(DBG_TLS, "received cipher suites unacceptable");
+				return FALSE;
+			}
+			this->server_auth->destroy(this->server_auth);
+			this->server_auth = auth_cfg_create();
+			key = lib->credmgr->get_private(lib->credmgr, type, this->server,
+											this->server_auth);
+			if (!key)
+			{
+				DBG1(DBG_TLS, "received cipher suites unacceptable");
+				return FALSE;
+			}
+		}
+	}
+	this->private = key;
+	return TRUE;
+}
+
+/**
+ * Check if the peer supports a given TLS curve
+ */
+static bool peer_supports_curve(private_tls_server_t *this,
+								tls_named_group_t curve)
+{
+	bio_reader_t *reader;
+	uint16_t current;
+
+	if (!this->curves_received)
+	{	/* none received, assume yes */
+		return TRUE;
+	}
+	reader = bio_reader_create(this->curves);
+	while (reader->remaining(reader) && reader->read_uint16(reader, &current))
+	{
+		if (current == curve)
 		{
-			DBG1(DBG_TLS, "received cipher suites unacceptable");
-			return FALSE;
+			reader->destroy(reader);
+			return TRUE;
 		}
-		this->server_auth->destroy(this->server_auth);
-		this->server_auth = auth_cfg_create();
-		key = lib->credmgr->get_private(lib->credmgr, type, this->server,
-										this->server_auth);
-		if (!key)
+	}
+	reader->destroy(reader);
+	return FALSE;
+}
+
+/**
+ * Try to find a curve supported by both, client and server
+ */
+static bool find_supported_curve(private_tls_server_t *this,
```

Comment:
Technically, you don't have to move this function as you don't call it in `process_client_hello()` anymore.